### PR TITLE
Client and server

### PR DESCRIPTION
**Pull request by andres-teleport**
Please review.

### CODE REVIEW

File: api/client/client.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+		if chunk != nil {
+			if _, err := w.Write(chunk.Output); err != nil {
+				w.CloseWithError(err)
+				return
+			}
+		}
+	}
+
+	w.Close()
```
Comment (smallinsky) [inline api/client/client.go]: missing err check.

File: api/server/server.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+func (s *Server) Stop(ctx context.Context, jobID *api.JobID) (*api.StopResponse, error) {
+	return &api.StopResponse{}, s.supervisor.StopJob(jobID.Id)
+}
+
+func (s *Server) Status(context context.Context, jobID *api.JobID) (*api.StatusResponse, error) {
+	st, err := s.supervisor.JobStatus(jobID.Id)
+	if err != nil {
+		return nil, err
```
Comment (smallinsky) [inline api/server/server.go]: err is raw go error thus by default it will be mapped to gRPC UNKNOWN status. 
Why only [ErrPermissionDenied](https://github.com/andres-teleport/overseer/pull/3/files#diff-8142f9f9e2f756f054e632b3fe796dfb9e9afc47529487934014ecf136101aadR14) grpc status is used ?

File: api/server/server.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+	"github.com/andres-teleport/overseer/api"
+	"github.com/andres-teleport/overseer/api/authentication"
+	"github.com/andres-teleport/overseer/lib/multipipe"
+	"github.com/andres-teleport/overseer/lib/supervisor"
+	"google.golang.org/grpc"
+)
+
+type Server struct {
+	jobOwners  map[string]string
```
Comment (smallinsky) [inline api/server/server.go:23]: Seems that user job information can be moved to `supervisor.Job` struct. Any reason why separate map was used ?

File: api/server/server.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+func (s *Server) Start(ctx context.Context, job *api.Job) (*api.JobID, error) {
+	jobID, err := s.supervisor.StartJob(job.Command, job.Arguments...)
+	if err != nil {
+		return nil, err
+	}
+
+	commonName, err := authentication.GetCommonNameFromCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
```
Comment (smallinsky) [inline api/server/server.go]: what if job was started  but `GetCommonNameFromCtx` call failed ? In that case started job is dangling. Changing the order for `StartJob` and `GetCommonNameFromCtx` calls should prevent such situations.

File: api/server/server.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+func (s *Server) Serve() error {
+	return s.srv.Serve(s.l)
+}
+
+func (s *Server) Start(ctx context.Context, job *api.Job) (*api.JobID, error) {
+	jobID, err := s.supervisor.StartJob(job.Command, job.Arguments...)
+	if err != nil {
+		return nil, err
```
Comment (smallinsky) [inline api/server/server.go:72]: missing gRPC Status code.

File: api/client/client.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		chunk, err := streamer.Recv()
+		if err == io.EOF {
+			eof = true
+		} else if err != nil {
+			w.CloseWithError(err)
+		}
+
+		if chunk != nil {
+			if _, err := w.Write(chunk.Output); err != nil {
+				w.CloseWithError(err)
```
Comment (smallinsky) [inline api/client/client.go]: ignored returned err

File: api/client/client.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	}, nil
+}
+
+func copyStream(streamer api.JobworkerService_StdOutClient, w *io.PipeWriter) {
+	for eof := false; !eof; {
+		chunk, err := streamer.Recv()
+		if err == io.EOF {
+			eof = true
+		} else if err != nil {
+			w.CloseWithError(err)
```
Comment (smallinsky) [inline api/client/client.go]: ignored returned err

File: api/server/server.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	return &api.StatusResponse{
+		Status:   status,
+		ExitCode: int64(st.ExitCode),
+	}, nil
+}
+
+func stream(jobID *api.JobID, srv grpc.ServerStream, sendFn func(*api.OutputChunk) error, fn func(string) (*multipipe.Reader, error)) error {
+	stdout, err := fn(jobID.Id)
```
Comment (smallinsky) [inline api/server/server.go]: `stdout` name is bit misleading because it can be also `StdErr`  reader

File: api/server/server.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+	if err != nil {
+		return err
```
Comment (smallinsky) [inline api/server/server.go]: It would useful to propagate context of the error.

File: api/server/server_test.go [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+		cli, err := u.newClientFn(getServerAddress(srv.l))
+		assertNil(t, err)
+
+		jobID, err := cli.Start("echo", "hello")
+		assertNil(t, err)
+
+		srv.mu.Lock()
+		if owner, ok := srv.jobOwners[jobID]; !ok {
+			srv.mu.Unlock()
+			t.FailNow()
```
Comment (smallinsky) [inline api/server/server_test.go]: failing without context can make test hard to debug so why not just use `t.Fatalf("job owner not found  for jobID:% v", jobID)`

File: api/server/server.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+	return s, nil
+}
+
+func (s *Server) Serve() error {
+	return s.srv.Serve(s.l)
+}
+
+func (s *Server) Start(ctx context.Context, job *api.Job) (*api.JobID, error) {
+	jobID, err := s.supervisor.StartJob(job.Command, job.Arguments...)
```
Comment (smallinsky) [inline api/server/server.go]: it would be nice to pre-check correctness of  user data and for example if job.Command is an empty string there is no need to call `supervisor.StartJob`

File: cmd/cli/main.go [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+		if status, err = cli.Status(statusJobID); err == nil {
+			if status.Status != "STARTED" {
+				fmt.Println(status.Status, "=", status.ExitCode)
+			} else {
+				fmt.Println(status.Status)
+			}
+		}
+	case len(stdOutJobID) > 0:
+		var rd *io.PipeReader
+		rd, err = cli.StdOut(stdOutJobID)
```
Comment (smallinsky) [inline cmd/cli/main.go]: missing `rd.Close` call.

File: cmd/cli/main.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+		if err == nil {
+			_, err = io.Copy(os.Stdout, rd)
+		}
+	case len(stdErrJobID) > 0:
+		var rd *io.PipeReader
+		rd, err = cli.StdErr(stdErrJobID)
```
Comment (smallinsky) [inline cmd/cli/main.go]: missing rd.Close call.

File: api/client/client.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+	"github.com/andres-teleport/overseer/api/authentication"
+	"google.golang.org/grpc"
+)
+
+type Client struct {
+	client api.JobworkerServiceClient
+}
+
+type Status struct {
+	Status   string
```
Comment (smallinsky) [inline api/client/client.go]: it will be more typo error prone to uses api.Status type here  instead of raw strings.

File: api/client/client.go [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+	client, err := c.client.StdOut(context.Background(), &api.JobID{Id: jobID})
+	if err != nil {
+		return nil, err
+	}
+
+	pr, pw := io.Pipe()
+	go copyStream(client, pw)
+
+	return pr, nil
+}
```
Comment (smallinsky) [inline api/client/client.go]: How do you handle situation when user of the library want to stop ongoing log stream ?

File: api/client/client.go (hunk 1 above)
Comment (andres-teleport) [reply api/client/client.go]: `w.Close()` cannot fail, it always returns `nil`.

File: api/server/server.go (hunk 3 above)
Comment (andres-teleport) [reply api/server/server.go:23]: I prefer the `lib/supervisor` package to be more general, and leave to the authorization part to upper layers.

File: api/client/client.go (hunk 6 above)
Comment (andres-teleport) [reply api/client/client.go]: `w.CloseWithError()` cannot fail, it always returns `nil`.

File: api/client/client.go (hunk 7 above)
Comment (andres-teleport) [reply api/client/client.go]: `w.CloseWithError()` cannot fail, it always returns `nil`.

File: api/client/client.go (hunk 14 above)
Comment (andres-teleport) [reply api/client/client.go]: Not sure I understand, are you saying that I should use `api.Status` directly instead of creating this new type?

Comment (smallinsky) [reply api/client/client.go]: yes, the `StatusResponse` gRPC message  uses enum type for status field,  for the same reason client library should also operate on enum status value instead of strings.

File: api/client/client.go (hunk 15 above)
Comment (andres-teleport) [reply api/client/client.go]: Perhaps I'm not understanding correctly but the reader could just close the returned `*io.PipeReader` if no more data needs to be read, this will make `w.Write(chunk.Output)` fail in `copyStream` and so it will close the writing end of the pipe.

File: api/server/server.go (hunk 2 above)
Comment (andres-teleport) [reply api/server/server.go]: Although I'm handling it, this error should never occur as the authorization interceptor will return a permission denied first, I'll mark it as internal so it is more clear.

File: api/client/client.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+	return &Client{client: api.NewJobworkerServiceClient(conn)}, nil
+}
+
+func (c *Client) Start(command string, arguments ...string) (string, error) {
+	job := &api.Job{
+		Command:   command,
+		Arguments: arguments,
+	}
+
+	jobID, err := c.client.Start(context.Background(), job)
```
Comment (r0mant) [inline api/client/client.go]: Nit: I would consider passing context to all Client methods.

File: api/client/client.go [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+	_, err := c.client.Stop(context.Background(), &api.JobID{Id: jobID})
+	return err
+}
+
+func (c *Client) Status(jobID string) (*api.StatusResponse, error) {
+	return c.client.Status(context.Background(), &api.JobID{Id: jobID})
+}
+
+func copyStream(streamer api.JobworkerService_StdOutClient, w *io.PipeWriter) {
+	for eof := false; !eof; {
```
Comment (r0mant) [inline api/client/client.go:54]: Nit: Might be a bit cleaner to modify the condition to `for err != io.EOF`, wouldn't need additional variable.

File: api/client/client.go [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+		if chunk != nil {
+			if _, err := w.Write(chunk.Output); err != nil {
+				_ = w.CloseWithError(err)
+				return
+			}
+		}
+	}
+
+	_ = w.Close()
```
Comment (r0mant) [inline api/client/client.go:70]: Are these errors from CloseXXX methods ignored on purpose?

File: api/client/client.go (hunk 15 above)
Comment (r0mant) [reply api/client/client.go]: Wouldn't the goroutine leak if the client is stuck on `streamer.Recv()` though?

File: api/server/server.go [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+	authInterceptor := NewAuthorizationInterceptor(s)
+
+	s.srv = grpc.NewServer(
+		grpc.Creds(creds),
+		grpc.UnaryInterceptor(authInterceptor.unaryServerInterceptor),
+		grpc.StreamInterceptor(authInterceptor.streamServerInterceptor),
+	)
+	api.RegisterJobworkerServiceServer(s.srv, s)
+
+	l, err := net.Listen("tcp", listenAddr)
```
Comment (r0mant) [inline api/server/server.go:52]: Might be useful to add a some sort of cleanup to the server which will close the listener.

File: api/server/server.go [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func stream(jobID *api.JobID, srv grpc.ServerStream, sendFn func(*api.OutputChunk) error, fn func(string) (*multipipe.Reader, error)) error {
+	out, err := fn(jobID.Id)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+
+	buf := make([]byte, 8192)
+	for eof := false; !eof; {
```
Comment (r0mant) [inline api/server/server.go:134]: Nit: Similar to my comment above, this probably could be reduced to `for err != io.EOF` to get rid of extra variable.

File: api/client/client.go (hunk 18 above)
Comment (andres-teleport) [reply api/client/client.go:70]: Yes, I added  the `_` to make it clear and to avoid noise with static checks, closing an `io.PipeWriter` will always return `nil`.

File: api/client/client.go (hunk 17 above)
Comment (andres-teleport) [reply api/client/client.go:54]: I tried but I'm not convinced, it's roughly the same number of lines but I have to be extra careful to avoid overwriting the error by mistake, so it seems more error-prone.

It looks like this:
```go
	var err error
	var chunk *api.OutputChunk
	for err != io.EOF {
		chunk, err = streamer.Recv()
		if err != nil && err != io.EOF {
			_ = w.CloseWithError(err)
		}

		if chunk != nil {
			if _, err := w.Write(chunk.Output); err != nil {
				_ = w.CloseWithError(err)
				return
			}
		}
	}
```

Please let me know if I'm missing anything.

File: api/server/server.go (hunk 20 above)
Comment (andres-teleport) [reply api/server/server.go:134]: See https://github.com/andres-teleport/overseer/pull/3#discussion_r686343412.

File: api/client/client.go (hunk 15 above)
Comment (andres-teleport) [reply api/client/client.go]: It won't really leak, but it will be stay around for longer than strictly necessary, that is, until the job finishes.

File: api/authentication/authentication_test.go [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+		{"/", "test-assets/server.crt", "test-assets/ca.crt"},
+
+		// Invalid cert
+		{"test-assets/server.key", "authentication_test.go", "test-assets/ca.crt"},
+		{"test-assets/server.key", "/", "test-assets/ca.crt"},
+
+		// Invalid CA
+		{"test-assets/server.key", "test-assets/server.crt", "authentication_test.go"},
+		{"test-assets/server.key", "test-assets/server.crt", "/"},
+	}
```
Comment (smallinsky) [inline api/authentication/authentication_test.go]: It hard to follow where args goes and in case of error it will be hard to debug with case failed. Test table should improve that:

```suggestion
	cs := []struct{
		desc string
		keyFile string
		certFile string
		caFile string
	} {
	    {
	        desc: "Invalid Key",
	        ...
	     }, 
	     {
	        ...
	     }
	}
	for _, tc := range cs {
		t.Run(tc.desc, func(t *testing.T){ 
		        ...
		}
	}
```

File: api/server/server.go [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+func (s *Server) Stop(ctx context.Context, jobID *api.JobID) (*api.StopResponse, error) {
+	err := s.supervisor.StopJob(jobID.Id)
+
+	if err == supervisor.ErrJobFinished {
+		err = status.Error(codes.FailedPrecondition, err.Error())
+	} else {
+		err = status.Error(codes.Internal, err.Error())
+	}
```
Comment (smallinsky) [inline api/server/server.go]: if err == nil the else statment will be invoked thus err == nil the `err.Error()` call in `err = status.Error(codes.Internal, err.Error())` will panic.
